runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3498,6 +3303,634 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+   call->setNeedsGCMap(0x0000FFFF);
+   call->setDependencyConditions(dependencies);
+
+   // After returning from the VM Helper, branch back to mainline code.
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, mergePointLabel);
+   // End of OOL.
+   outlinedSlowPath->swapInstructionListsWithCompilation();
+
+   cg->stopUsingRegister(scratch1);
+   cg->stopUsingRegister(scratch2);
+
+   cg->stopUsingRegister(dataBlockReg);
+   }
+
+void
+J9::Z::TreeEvaluator::readWriteBarrierHelperForFieldWatch(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister, TR::Register *valueReg)
+   {
+   bool isResolved = !node->getSymbolReference()->isUnresolved();
+   TR::LabelSymbol *mergePointLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *fieldReportLabel = generateLabelSymbol(cg);
+
+   TR::Register *fieldClassReg;
+   TR::Register *fieldClassFlags = cg->allocateRegister();
+   bool opCodeIsIndirect = node->getOpCode().isIndirect();
+
+   if (opCodeIsIndirect)
+      {
+      // Load the class of the instance object into fieldClassReg.
+      fieldClassReg = cg->allocateRegister();
+      TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, fieldClassReg, generateS390MemoryReference(sideEffectRegister, static_cast<int32_t>(TR::Compiler->om.offsetOfObjectVftField()), cg), NULL);
+      }
+   else
+      {
+      if (isResolved)
+         {
+         fieldClassReg = cg->allocateRegister();
+         if (!(cg->comp()->compileRelocatableCode()))
+            {
+            // For non-AOT compiles we don't need to use sideEffectRegister here as the class information is available to us at compile time.
+            J9Class *fieldClass = static_cast<TR::J9WatchedStaticFieldSnippet *>(dataSnippet)->getFieldClass();
+            TR_ASSERT_FATAL(fieldClass != NULL, "A valid J9Class must be provided for direct rdbar/wrtbar opcodes %p\n", node);
+            generateRILInstruction(cg, TR::InstOpCode::LARL, node, fieldClassReg, static_cast<void *>(fieldClass));
+            }
+         else
+            {
+            // If this is an AOT compile, we generate instructions to load the fieldClass directly from the snippet because the fieldClass will be invalid
+            // if we load using the dataSnippet's helper query at compile time.
+            generateRILInstruction(cg, TR::InstOpCode::LARL, node, fieldClassReg, dataSnippet);
+            generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, fieldClassReg, generateS390MemoryReference(fieldClassReg, offsetof(J9JITWatchedStaticFieldData, fieldClass), cg));
+            }
+         }
+      else
+         {
+         if (isWrite)
+            {
+            fieldClassReg = cg->allocateRegister();
+            generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, fieldClassReg, generateS390MemoryReference(sideEffectRegister, cg->comp()->fej9()->getOffsetOfClassFromJavaLangClassField(), cg));
```

Comment:
@gita-omr FYI we load the J9Class from a static `wrtbar`'s second child directly with this instruction. A similar one exists on X.